File: packages/snow_draw_core/lib/draw/elements/types/shared/hit_test_geometry.dart
import 'dart:math' as math;

import '../../../core/coordinates/element_space.dart';
import '../../../models/element_state.dart';
import '../../../types/draw_point.dart';
import '../../../types/draw_rect.dart';

/// Converts [position] into the element-local coordinate space.
DrawPoint resolveElementLocalPosition({
  required ElementState element,
  required DrawPoint position,
}) {
  if (element.rotation == 0) {
    return position;
  }
  final rect = element.rect;
  final space = ElementSpace(rotation: element.rotation, origin: rect.center);
  return space.fromWorld(position);
}

/// Returns whether [value] has finite coordinates.
bool isFiniteDrawPoint(DrawPoint value) => value.x.isFinite && value.y.isFinite;

/// Returns whether [position] is inside [rect] expanded by [padding].
bool isPointInsideRect(DrawRect rect, DrawPoint position, double padding) =>
    position.x >= rect.minX - padding &&
    position.x <= rect.maxX + padding &&
    position.y >= rect.minY - padding &&
    position.y <= rect.maxY + padding;

/// Hit-tests a normalized two-point stroke against [localPosition].
///
/// [normalizedStart] and [normalizedEnd] are expected in [0, 1] element space.
bool hitTestNormalizedTwoPointStroke({
  required DrawRect rect,
  required DrawPoint normalizedStart,
  required DrawPoint normalizedEnd,
  required DrawPoint localPosition,
  required double strokeWidth,
  required double tolerance,
}) {
  final radius = (strokeWidth / 2) + tolerance;
  if (!radius.isFinite || radius <= 0) {
    return false;
  }
  if (!isPointInsideRect(rect, localPosition, radius)) {
    return false;
  }
  if (!rect.width.isFinite ||
      !rect.height.isFinite ||
      rect.width < 0 ||
      rect.height < 0) {
    return false;
  }

  final start = DrawPoint(
    x: rect.minX + (normalizedStart.x * rect.width),
    y: rect.minY + (normalizedStart.y * rect.height),
  );
  final end = DrawPoint(
    x: rect.minX + (normalizedEnd.x * rect.width),
    y: rect.minY + (normalizedEnd.y * rect.height),
  );
  if (!isFiniteDrawPoint(start) || !isFiniteDrawPoint(end)) {
    return false;
  }
  return distanceSquaredToSegment(localPosition, start, end) <= radius * radius;
}

/// Returns squared distance from point [p] to segment [a]-[b].
double distanceSquaredToSegment(DrawPoint p, DrawPoint a, DrawPoint b) {
  final abX = b.x - a.x;
  final abY = b.y - a.y;
  final apX = p.x - a.x;
  final apY = p.y - a.y;
  final abLengthSq = abX * abX + abY * abY;
  if (abLengthSq == 0) {
    final dx = apX;
    final dy = apY;
    return dx * dx + dy * dy;
  }
  var t = (apX * abX + apY * abY) / abLengthSq;
  if (t < 0) {
    t = 0;
  } else if (t > 1) {
    t = 1;
  }
  final closestX = a.x + abX * t;
  final closestY = a.y + abY * t;
  final dx = p.x - closestX;
  final dy = p.y - closestY;
  return dx * dx + dy * dy;
}

/// Returns squared distance from [point] to the infinite line through [a]-[b].
double _distanceSquaredToLine(DrawPoint point, DrawPoint a, DrawPoint b) {
  final dx = b.x - a.x;
  final dy = b.y - a.y;
  final lenSq = dx * dx + dy * dy;
  if (lenSq == 0) {
    final diffX = point.x - a.x;
    final diffY = point.y - a.y;
    return diffX * diffX + diffY * diffY;
  }
  final cross = dx * (point.y - a.y) - dy * (point.x - a.x);
  return (cross * cross) / lenSq;
}

/// Returns true when [point] lies inside [polygon] using even-odd winding.
bool isPointInsidePolygon(
  DrawPoint point,
  List<DrawPoint> polygon, {
  double epsilon = 1e-9,
}) {
  if (polygon.length < 3) {
    return false;
  }

  var inside = false;
  for (var i = 0, j = polygon.length - 1; i < polygon.length; j = i++) {
    final current = polygon[i];
    final previous = polygon[j];
    final intersects =
        (current.y > point.y) != (previous.y > point.y) &&
        point.x <
            (previous.x - current.x) *
                    (point.y - current.y) /
                    ((previous.y - current.y).abs() < epsilon
                        ? epsilon
                        : previous.y - current.y) +
                current.x;
    if (intersects) {
      inside = !inside;
    }
  }
  return inside;
}

/// Returns [a] + [b].
DrawPoint addDrawPoints(DrawPoint a, DrawPoint b) =>
    DrawPoint(x: a.x + b.x, y: a.y + b.y);

/// Returns [a] + [direction] * [scale].
DrawPoint addScaledDrawPoint(DrawPoint a, DrawPoint direction, double scale) =>
    DrawPoint(x: a.x + direction.x * scale, y: a.y + direction.y * scale);

/// Returns [a] - [direction] * [scale].
DrawPoint subtractScaledDrawPoint(
  DrawPoint a,
  DrawPoint direction,
  double scale,
) => DrawPoint(x: a.x - direction.x * scale, y: a.y - direction.y * scale);

/// Returns the midpoint of [a] and [b].
DrawPoint _midpointDrawPoint(DrawPoint a, DrawPoint b) =>
    DrawPoint(x: (a.x + b.x) * 0.5, y: (a.y + b.y) * 0.5);

/// Returns normalized [value], or null when it has zero length.
DrawPoint? normalizeDrawVector(DrawPoint value) {
  final length = math.sqrt(value.x * value.x + value.y * value.y);
  if (length == 0) {
    return null;
  }
  return DrawPoint(x: value.x / length, y: value.y / length);
}

/// Cubic segment used for curve flattening.
class CubicDrawSegment {
  /// Creates an immutable cubic segment.
  const CubicDrawSegment({
    required this.start,
    required this.control1,
    required this.control2,
    required this.end,
  });

  /// Segment start point.
  final DrawPoint start;

  /// First control point.
  final DrawPoint control1;

  /// Second control point.
  final DrawPoint control2;

  /// Segment end point.
  final DrawPoint end;
}

/// Builds a Catmull-Rom-derived cubic segment at [index].
CubicDrawSegment buildCatmullRomCubicSegment(
  List<DrawPoint> points,
  int index, {
  double tension = 1.0,
}) {
  final p0 = index == 0 ? points[index] : points[index - 1];
  final p1 = points[index];
  final p2 = points[index + 1];
  final p3 = index + 2 < points.length ? points[index + 2] : points[index + 1];

  final control1 = DrawPoint(
    x: p1.x + (p2.x - p0.x) * (tension / 6),
    y: p1.y + (p2.y - p0.y) * (tension / 6),
  );
  final control2 = DrawPoint(
    x: p2.x - (p3.x - p1.x) * (tension / 6),
    y: p2.y - (p3.y - p1.y) * (tension / 6),
  );
  return CubicDrawSegment(
    start: p1,
    control1: control1,
    control2: control2,
    end: p2,
  );
}

/// Flattens Catmull-Rom points into a polyline for hit testing.
List<DrawPoint> flattenCatmullRomDrawPoints({
  required List<DrawPoint> points,
  required double strokeWidth,
  int maxPoints = 120,
}) {
  if (points.isEmpty) {
    return const <DrawPoint>[];
  }
  if (points.length < 3) {
    return points;
  }

  final step = math.max(1, strokeWidth).toDouble();
  final tolerance = math.max(0.5, step * 0.35);
  final toleranceSq = tolerance * tolerance;

  final flattened = <DrawPoint>[points.first];
  for (var i = 0; i < points.length - 1; i++) {
    if (flattened.length >= maxPoints) {
      break;
    }
    final segment = buildCatmullRomCubicSegment(points, i);
    _flattenCubicSegment(
      segment: segment,
      toleranceSq: toleranceSq,
      output: flattened,
      maxPoints: maxPoints,
    );
  }

  return flattened;
}

void _flattenCubicSegment({
  required CubicDrawSegment segment,
  required double toleranceSq,
  required List<DrawPoint> output,
  required int maxPoints,
}) {
  final stack = <CubicDrawSegment>[segment];
  while (stack.isNotEmpty && output.length < maxPoints) {
    final current = stack.removeLast();
    if (_isCubicFlatEnough(current, toleranceSq) ||
        output.length >= maxPoints - 1) {
      output.add(current.end);
      continue;
    }

    final split = _splitCubicSegment(current);
    stack
      ..add(split.right)
      ..add(split.left);
  }
}

bool _isCubicFlatEnough(CubicDrawSegment segment, double toleranceSq) {
  final dist1 = _distanceSquaredToLine(
    segment.control1,
    segment.start,
    segment.end,
  );
  final dist2 = _distanceSquaredToLine(
    segment.control2,
    segment.start,
    segment.end,
  );
  return math.max(dist1, dist2) <= toleranceSq;
}

({CubicDrawSegment left, CubicDrawSegment right}) _splitCubicSegment(
  CubicDrawSegment segment,
) {
  final p01 = _midpointDrawPoint(segment.start, segment.control1);
  final p12 = _midpointDrawPoint(segment.control1, segment.control2);
  final p23 = _midpointDrawPoint(segment.control2, segment.end);
  final p012 = _midpointDrawPoint(p01, p12);
  final p123 = _midpointDrawPoint(p12, p23);
  final p0123 = _midpointDrawPoint(p012, p123);

  return (
    left: CubicDrawSegment(
      start: segment.start,
      control1: p01,
      control2: p012,
      end: p0123,
    ),
    right: CubicDrawSegment(
      start: p0123,
      control1: p123,
      control2: p23,
      end: segment.end,
    ),
  );
}

File: packages/snow_draw_rust_engine/src/draw/elements/types/shared/hit_test_geometry.rs
#![allow(dead_code)]

use crate::draw::core::coordinates::element_space::ElementSpace;
use crate::draw::types::draw_point::DrawPoint;
use crate::draw::types::draw_rect::DrawRect;

/// Converts `position` into the element-local coordinate space.
///
/// `rect` and `rotation` are the element's world-space bounds and rotation.
pub fn resolve_element_local_position(
    rect: DrawRect,
    rotation: f64,
    position: DrawPoint,
) -> DrawPoint {
    if rotation == 0.0 {
        return position;
    }
    let space = ElementSpace::new(rotation, rect.center());
    space.from_world(position)
}

/// Returns whether `value` has finite coordinates.
pub fn is_finite_draw_point(value: DrawPoint) -> bool {
    value.x.is_finite() && value.y.is_finite()
}

/// Returns whether `position` is inside `rect` expanded by `padding`.
pub fn is_point_inside_rect(rect: DrawRect, position: DrawPoint, padding: f64) -> bool {
    position.x >= rect.min_x - padding
        && position.x <= rect.max_x + padding
        && position.y >= rect.min_y - padding
        && position.y <= rect.max_y + padding
}

/// Hit-tests a normalized two-point stroke against `local_position`.
///
/// `normalized_start` and `normalized_end` are expected in `[0, 1]`
/// element-space coordinates.
pub fn hit_test_normalized_two_point_stroke(
    rect: DrawRect,
    normalized_start: DrawPoint,
    normalized_end: DrawPoint,
    local_position: DrawPoint,
    stroke_width: f64,
    tolerance: f64,
) -> bool {
    let radius = (stroke_width / 2.0) + tolerance;
    if !radius.is_finite() || radius <= 0.0 {
        return false;
    }
    if !is_point_inside_rect(rect, local_position, radius) {
        return false;
    }

    let width = rect.width();
    let height = rect.height();
    if !width.is_finite() || !height.is_finite() || width < 0.0 || height < 0.0 {
        return false;
    }

    let start = DrawPoint::new(
        rect.min_x + (normalized_start.x * width),
        rect.min_y + (normalized_start.y * height),
    );
    let end = DrawPoint::new(
        rect.min_x + (normalized_end.x * width),
        rect.min_y + (normalized_end.y * height),
    );
    if !is_finite_draw_point(start) || !is_finite_draw_point(end) {
        return false;
    }

    distance_squared_to_segment(local_position, start, end) <= radius * radius
}

/// Returns squared distance from point `p` to segment `a-b`.
pub fn distance_squared_to_segment(p: DrawPoint, a: DrawPoint, b: DrawPoint) -> f64 {
    let ab_x = b.x - a.x;
    let ab_y = b.y - a.y;
    let ap_x = p.x - a.x;
    let ap_y = p.y - a.y;
    let ab_length_sq = ab_x * ab_x + ab_y * ab_y;

    if ab_length_sq == 0.0 {
        return ap_x * ap_x + ap_y * ap_y;
    }

    let mut t = (ap_x * ab_x + ap_y * ab_y) / ab_length_sq;
    t = t.clamp(0.0, 1.0);

    let closest_x = a.x + ab_x * t;
    let closest_y = a.y + ab_y * t;
    let dx = p.x - closest_x;
    let dy = p.y - closest_y;
    dx * dx + dy * dy
}

/// Returns true when `point` lies inside `polygon` using even-odd winding.
pub fn is_point_inside_polygon(point: DrawPoint, polygon: &[DrawPoint], epsilon: f64) -> bool {
    if polygon.len() < 3 {
        return false;
    }

    let mut inside = false;
    let mut previous_index = polygon.len() - 1;
    for (index, current) in polygon.iter().copied().enumerate() {
        let previous = polygon[previous_index];
        let intersects = (current.y > point.y) != (previous.y > point.y)
            && point.x
                < (previous.x - current.x) * (point.y - current.y)
                    / if (previous.y - current.y).abs() < epsilon {
                        epsilon
                    } else {
                        previous.y - current.y
                    }
                    + current.x;
        if intersects {
            inside = !inside;
        }
        previous_index = index;
    }
    inside
}

/// Returns `a + b`.
pub fn add_draw_points(a: DrawPoint, b: DrawPoint) -> DrawPoint {
    DrawPoint::new(a.x + b.x, a.y + b.y)
}

/// Returns `a + direction * scale`.
pub fn add_scaled_draw_point(a: DrawPoint, direction: DrawPoint, scale: f64) -> DrawPoint {
    DrawPoint::new(a.x + direction.x * scale, a.y + direction.y * scale)
}

/// Returns `a - direction * scale`.
pub fn subtract_scaled_draw_point(a: DrawPoint, direction: DrawPoint, scale: f64) -> DrawPoint {
    DrawPoint::new(a.x - direction.x * scale, a.y - direction.y * scale)
}

/// Returns normalized `value`, or `None` when it has zero length.
pub fn normalize_draw_vector(value: DrawPoint) -> Option<DrawPoint> {
    let length = (value.x * value.x + value.y * value.y).sqrt();
    if length == 0.0 {
        return None;
    }
    Some(DrawPoint::new(value.x / length, value.y / length))
}

/// Cubic segment used for curve flattening.
#[derive(Clone, Copy, Debug, PartialEq)]
pub struct CubicDrawSegment {
    /// Segment start point.
    pub start: DrawPoint,
    /// First control point.
    pub control1: DrawPoint,
    /// Second control point.
    pub control2: DrawPoint,
    /// Segment end point.
    pub end: DrawPoint,
}

impl CubicDrawSegment {
    /// Creates an immutable cubic segment.
    pub const fn new(
        start: DrawPoint,
        control1: DrawPoint,
        control2: DrawPoint,
        end: DrawPoint,
    ) -> Self {
        Self {
            start,
            control1,
            control2,
            end,
        }
    }
}

/// Builds a Catmull-Rom-derived cubic segment at `index`.
pub fn build_catmull_rom_cubic_segment(
    points: &[DrawPoint],
    index: usize,
    tension: f64,
    phantom_first: Option<DrawPoint>,
    phantom_last: Option<DrawPoint>,
) -> CubicDrawSegment {
    let last_index = points.len() - 1;
    let p0 = if index == 0 {
        phantom_first.unwrap_or(points[index])
    } else {
        points[index - 1]
    };
    let p1 = points[index];
    let p2 = points[index + 1];
    let p3 = if index + 2 <= last_index {
        points[index + 2]
    } else {
        phantom_last.unwrap_or(points[index + 1])
    };

    let control1 = DrawPoint::new(
        p1.x + (p2.x - p0.x) * (tension / 6.0),
        p1.y + (p2.y - p0.y) * (tension / 6.0),
    );
    let control2 = DrawPoint::new(
        p2.x - (p3.x - p1.x) * (tension / 6.0),
        p2.y - (p3.y - p1.y) * (tension / 6.0),
    );

    CubicDrawSegment::new(p1, control1, control2, p2)
}

/// Flattens Catmull-Rom points into a polyline for hit testing.
pub fn flatten_catmull_rom_draw_points(
    points: &[DrawPoint],
    stroke_width: f64,
    max_points: usize,
    tension: f64,
    use_open_endpoint_phantom_points: bool,
) -> Vec<DrawPoint> {
    if points.is_empty() {
        return Vec::new();
    }
    if points.len() < 3 {
        return points.to_vec();
    }
    if max_points == 0 {
        return Vec::new();
    }

    let step = stroke_width.max(1.0);
    let tolerance = 0.5_f64.max(step * 0.35);
    let tolerance_sq = tolerance * tolerance;

    let mut phantom_first: Option<DrawPoint> = None;
    let mut phantom_last: Option<DrawPoint> = None;
    if use_open_endpoint_phantom_points && points.len() >= 2 {
        let first = points[0];
        let second = points[1];
        let penultimate = points[points.len() - 2];
        let last = points[points.len() - 1];
        phantom_first = Some(first + (first - second));
        phantom_last = Some(last + (last - penultimate));
    }

    let mut flattened = vec![points[0]];
    for index in 0..(points.len() - 1) {
        if flattened.len() >= max_points {
            break;
        }
        let segment =
            build_catmull_rom_cubic_segment(points, index, tension, phantom_first, phantom_last);
        flatten_cubic_segment(segment, tolerance_sq, &mut flattened, max_points);
    }
    flattened
}

fn flatten_cubic_segment(
    segment: CubicDrawSegment,
    tolerance_sq: f64,
    output: &mut Vec<DrawPoint>,
    max_points: usize,
) {
    if max_points == 0 {
        return;
    }

    let mut stack = vec![segment];
    while !stack.is_empty() && output.len() < max_points {
        let current = stack.pop().expect("stack is not empty");
        if is_cubic_flat_enough(current, tolerance_sq)
            || output.len() >= max_points.saturating_sub(1)
        {
            output.push(current.end);
            continue;
        }

        let (left, right) = split_cubic_segment(current);
        stack.push(right);
        stack.push(left);
    }
}

fn is_cubic_flat_enough(segment: CubicDrawSegment, tolerance_sq: f64) -> bool {
    let dist1 = distance_squared_to_line(segment.control1, segment.start, segment.end);
    let dist2 = distance_squared_to_line(segment.control2, segment.start, segment.end);
    dist1.max(dist2) <= tolerance_sq
}

fn split_cubic_segment(segment: CubicDrawSegment) -> (CubicDrawSegment, CubicDrawSegment) {
    let p01 = midpoint_draw_point(segment.start, segment.control1);
    let p12 = midpoint_draw_point(segment.control1, segment.control2);
    let p23 = midpoint_draw_point(segment.control2, segment.end);
    let p012 = midpoint_draw_point(p01, p12);
    let p123 = midpoint_draw_point(p12, p23);
    let p0123 = midpoint_draw_point(p012, p123);

    (
        CubicDrawSegment::new(segment.start, p01, p012, p0123),
        CubicDrawSegment::new(p0123, p123, p23, segment.end),
    )
}

fn midpoint_draw_point(a: DrawPoint, b: DrawPoint) -> DrawPoint {
    DrawPoint::new((a.x + b.x) * 0.5, (a.y + b.y) * 0.5)
}

fn distance_squared_to_line(point: DrawPoint, a: DrawPoint, b: DrawPoint) -> f64 {
    let dx = b.x - a.x;
    let dy = b.y - a.y;
    let len_sq = dx * dx + dy * dy;
    if len_sq == 0.0 {
        let diff_x = point.x - a.x;
        let diff_y = point.y - a.y;
        return diff_x * diff_x + diff_y * diff_y;
    }
    let cross = dx * (point.y - a.y) - dy * (point.x - a.x);
    (cross * cross) / len_sq
}
